File: exchange-core/src/main/java/exchange/core2/core/processors/loan/LoanGlobalConfig.java
/*
 * Copyright 2019 Maksim Zheravin
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package exchange.core2.core.processors.loan;

import java.util.Objects;

import exchange.core2.core.common.StateHash;
import net.openhft.chronicle.bytes.BytesIn;
import net.openhft.chronicle.bytes.BytesOut;
import net.openhft.chronicle.bytes.WriteBytesMarshallable;

/**
 * 现货借贷全局（per-shard 单例）运行时配置：Cross 强平线 / 预警线、池利用率上限、强平费率、numeraire。进 raft snapshot、跨 shard 独立， 仅 {@code ADD_LOAN} 命令逐字段
 * partial-update 改写（见 loan.md §11）。
 */
public final class LoanGlobalConfig implements WriteBytesMarshallable, StateHash {
    public static final int DEFAULT_CROSS_LIQUIDATION_LTV_BPS = 8500; // 85%
    public static final int DEFAULT_CROSS_MARGIN_CALL_LTV_BPS = 8000; // 80%
    public static final int DEFAULT_LOAN_POOL_UTILIZATION_CAP_BPS = 9000; // 90%
    public static final int DEFAULT_LOAN_LIQUIDATION_FEE_BPS = 200; // 2%
    public static final int DEFAULT_LTV_LIQUIDATION_BUFFER_BPS = 2000; // 20% initial→liquidation
    public static final int DEFAULT_LTV_MARGIN_CALL_BUFFER_BPS = 1000; // 10% liquidation→marginCall
    public static final int NUMERAIRE_UNSET = 0; // numeraire 未配置的 sentinel

    public int numeraireCurrency; // Cross 估值基准币；NUMERAIRE_UNSET(0)=未配 → Cross BORROW/WITHDRAW fail-close、scanner 跳过
    public int crossLiquidationLtvBps; // Cross 账户级强平线（bps）
    public int crossMarginCallLtvBps; // Cross 账户级预警线（bps）
    public int loanPoolUtilizationCapBps; // 借贷池利用率上限（bps）
    public int loanLiquidationFeeBps; // 强平专项费率（bps）
    public int ltvLiquidationBufferBps; // Symbol 派生:liquidation = initial + 本值
    public int ltvMarginCallBufferBps;  // Symbol/Cross 派生:marginCall = liquidation − 本值

    public LoanGlobalConfig() {
        this.numeraireCurrency = NUMERAIRE_UNSET;
        this.crossLiquidationLtvBps = DEFAULT_CROSS_LIQUIDATION_LTV_BPS;
        this.crossMarginCallLtvBps = DEFAULT_CROSS_MARGIN_CALL_LTV_BPS;
        this.loanPoolUtilizationCapBps = DEFAULT_LOAN_POOL_UTILIZATION_CAP_BPS;
        this.loanLiquidationFeeBps = DEFAULT_LOAN_LIQUIDATION_FEE_BPS;
        this.ltvLiquidationBufferBps = DEFAULT_LTV_LIQUIDATION_BUFFER_BPS;
        this.ltvMarginCallBufferBps = DEFAULT_LTV_MARGIN_CALL_BUFFER_BPS;
    }

    public LoanGlobalConfig(BytesIn bytes) {
        this.numeraireCurrency = bytes.readInt();
        this.crossLiquidationLtvBps = bytes.readInt();
        this.crossMarginCallLtvBps = bytes.readInt();
        this.loanPoolUtilizationCapBps = bytes.readInt();
        this.loanLiquidationFeeBps = bytes.readInt();
        this.ltvLiquidationBufferBps = bytes.readInt();
        this.ltvMarginCallBufferBps = bytes.readInt();
    }

    public void reset() {
        this.numeraireCurrency = NUMERAIRE_UNSET;
        this.crossLiquidationLtvBps = DEFAULT_CROSS_LIQUIDATION_LTV_BPS;
        this.crossMarginCallLtvBps = DEFAULT_CROSS_MARGIN_CALL_LTV_BPS;
        this.loanPoolUtilizationCapBps = DEFAULT_LOAN_POOL_UTILIZATION_CAP_BPS;
        this.loanLiquidationFeeBps = DEFAULT_LOAN_LIQUIDATION_FEE_BPS;
        this.ltvLiquidationBufferBps = DEFAULT_LTV_LIQUIDATION_BUFFER_BPS;
        this.ltvMarginCallBufferBps = DEFAULT_LTV_MARGIN_CALL_BUFFER_BPS;
    }

    /** numeraire 是否已配置（未配则 Cross 借贷 fail-close、scanner 跳过 Cross）。 */
    public boolean isNumeraireConfigured() {
        return numeraireCurrency != NUMERAIRE_UNSET;
    }

    @Override
    public void writeMarshallable(BytesOut bytes) {
        bytes.writeInt(numeraireCurrency);
        bytes.writeInt(crossLiquidationLtvBps);
        bytes.writeInt(crossMarginCallLtvBps);
        bytes.writeInt(loanPoolUtilizationCapBps);
        bytes.writeInt(loanLiquidationFeeBps);
        bytes.writeInt(ltvLiquidationBufferBps);
        bytes.writeInt(ltvMarginCallBufferBps);
    }

    @Override
    public int stateHash() {
        return Objects.hash(numeraireCurrency, crossLiquidationLtvBps, crossMarginCallLtvBps, loanPoolUtilizationCapBps,
            loanLiquidationFeeBps, ltvLiquidationBufferBps, ltvMarginCallBufferBps);
    }

    @Override
    public String toString() {
        return "LoanGlobalConfig{numeraire=" + numeraireCurrency + " crossLiqLtv=" + crossLiquidationLtvBps
            + " crossMcLtv=" + crossMarginCallLtvBps + " poolCap=" + loanPoolUtilizationCapBps + " liqFee="
            + loanLiquidationFeeBps + " liqBuf=" + ltvLiquidationBufferBps + " mcBuf=" + ltvMarginCallBufferBps + '}';
    }
}

File: exchange-core-rs/src/core/processors/loan/loan_global_config.rs
pub const DEFAULT_CROSS_LIQUIDATION_LTV_BPS: i32 = 8500;

pub const DEFAULT_CROSS_MARGIN_CALL_LTV_BPS: i32 = 8000;

pub const DEFAULT_LOAN_POOL_UTILIZATION_CAP_BPS: i32 = 9000;

pub const DEFAULT_LOAN_LIQUIDATION_FEE_BPS: i32 = 200;

pub const DEFAULT_LTV_LIQUIDATION_BUFFER_BPS: i32 = 2000;

pub const DEFAULT_LTV_MARGIN_CALL_BUFFER_BPS: i32 = 1000;

pub const NUMERAIRE_UNSET: i32 = 0;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LoanGlobalConfig {
    pub numeraire_currency: i32,
    pub cross_liquidation_ltv_bps: i32,
    pub cross_margin_call_ltv_bps: i32,
    pub loan_pool_utilization_cap_bps: i32,
    pub loan_liquidation_fee_bps: i32,
    pub ltv_liquidation_buffer_bps: i32,
    pub ltv_margin_call_buffer_bps: i32,
}

impl LoanGlobalConfig {

    pub fn reset(&mut self) {
        *self = LoanGlobalConfig::default();
    }

    pub fn is_numeraire_configured(&self) -> bool {
        self.numeraire_currency != NUMERAIRE_UNSET
    }

    pub fn state_hash(&self) -> i32 {
        let mut h: i64 = 17;
        h = h.wrapping_mul(31).wrapping_add(self.numeraire_currency as i64);
        h = h.wrapping_mul(31).wrapping_add(self.cross_liquidation_ltv_bps as i64);
        h = h.wrapping_mul(31).wrapping_add(self.cross_margin_call_ltv_bps as i64);
        h = h.wrapping_mul(31).wrapping_add(self.loan_pool_utilization_cap_bps as i64);
        h = h.wrapping_mul(31).wrapping_add(self.loan_liquidation_fee_bps as i64);
        h = h.wrapping_mul(31).wrapping_add(self.ltv_liquidation_buffer_bps as i64);
        h = h.wrapping_mul(31).wrapping_add(self.ltv_margin_call_buffer_bps as i64);
        ((h >> 32) as i32) ^ (h as i32)
    }
}

impl Default for LoanGlobalConfig {
    fn default() -> Self {
        LoanGlobalConfig {
            numeraire_currency: NUMERAIRE_UNSET,
            cross_liquidation_ltv_bps: DEFAULT_CROSS_LIQUIDATION_LTV_BPS,
            cross_margin_call_ltv_bps: DEFAULT_CROSS_MARGIN_CALL_LTV_BPS,
            loan_pool_utilization_cap_bps: DEFAULT_LOAN_POOL_UTILIZATION_CAP_BPS,
            loan_liquidation_fee_bps: DEFAULT_LOAN_LIQUIDATION_FEE_BPS,
            ltv_liquidation_buffer_bps: DEFAULT_LTV_LIQUIDATION_BUFFER_BPS,
            ltv_margin_call_buffer_bps: DEFAULT_LTV_MARGIN_CALL_BUFFER_BPS,
        }
    }
}

use crate::core::snapshot::chronicle_reader::{ChronicleError, ChronicleReader};
use crate::core::snapshot::chronicle_writer::ChronicleWriter;
use crate::core::snapshot::marshalling::ChronicleMarshallable;

impl ChronicleMarshallable for LoanGlobalConfig {

    fn chronicle_write(&self, w: &mut ChronicleWriter) {
        w.write_i32(self.numeraire_currency);
        w.write_i32(self.cross_liquidation_ltv_bps);
        w.write_i32(self.cross_margin_call_ltv_bps);
        w.write_i32(self.loan_pool_utilization_cap_bps);
        w.write_i32(self.loan_liquidation_fee_bps);
        w.write_i32(self.ltv_liquidation_buffer_bps);
        w.write_i32(self.ltv_margin_call_buffer_bps);
    }

    fn chronicle_read(r: &mut ChronicleReader) -> Result<Self, ChronicleError> {
        Ok(LoanGlobalConfig {
            numeraire_currency: r.read_i32()?,
            cross_liquidation_ltv_bps: r.read_i32()?,
            cross_margin_call_ltv_bps: r.read_i32()?,
            loan_pool_utilization_cap_bps: r.read_i32()?,
            loan_liquidation_fee_bps: r.read_i32()?,
            ltv_liquidation_buffer_bps: r.read_i32()?,
            ltv_margin_call_buffer_bps: r.read_i32()?,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_matches_java_seven_defaults() {
        let cfg = LoanGlobalConfig::default();
        assert_eq!(cfg.numeraire_currency, 0);
        assert_eq!(cfg.cross_liquidation_ltv_bps, 8500);
        assert_eq!(cfg.cross_margin_call_ltv_bps, 8000);
        assert_eq!(cfg.loan_pool_utilization_cap_bps, 9000);
        assert_eq!(cfg.loan_liquidation_fee_bps, 200);
        assert_eq!(cfg.ltv_liquidation_buffer_bps, 2000);
        assert_eq!(cfg.ltv_margin_call_buffer_bps, 1000);
    }

    #[test]
    fn is_numeraire_configured_tracks_unset_sentinel() {
        let mut cfg = LoanGlobalConfig::default();
        assert!(!cfg.is_numeraire_configured());
        cfg.numeraire_currency = 5;
        assert!(cfg.is_numeraire_configured());
        cfg.numeraire_currency = NUMERAIRE_UNSET;
        assert!(!cfg.is_numeraire_configured());
    }

    #[test]
    fn reset_restores_defaults_after_mutation() {
        let mut cfg = LoanGlobalConfig::default();
        cfg.numeraire_currency = 7;
        cfg.cross_liquidation_ltv_bps = 1;
        cfg.reset();
        assert_eq!(cfg, LoanGlobalConfig::default());
    }

    #[test]
    fn state_hash_deterministic_and_sensitive_to_field_change() {
        let a = LoanGlobalConfig::default();
        let b = LoanGlobalConfig::default();
        assert_eq!(a.state_hash(), b.state_hash());

        let mut c = a;
        c.numeraire_currency = 1;
        assert_ne!(a.state_hash(), c.state_hash());
    }
}
